lib/checknullpointer.cpp

```diff
@@ -509,31 +509,51 @@ void CheckNullPointer::arithmetic()
     for (std::size_t i = 0; i < functions; ++i) {
         const Scope * scope = symbolDatabase->functionScopes[i];
         for (const Token* tok = scope->classStart->next(); tok != scope->classEnd; tok = tok->next()) {
-            if (!tok->astOperand2() || tok->str() != "-")
+            if (!Token::Match(tok, "-|+|+=|-="))
                 continue;
-            // pointer subtraction
-            if (!tok->valueType() || !tok->valueType()->pointer)
-                continue;
-            // Can LHS be NULL?
-            const ValueFlow::Value *value = tok->astOperand1()->getValue(0);
-            if (!value)
-                continue;
-            if (!_settings->inconclusive && value->isInconclusive())
-                continue;
-            if (value->condition && !_settings->isEnabled(Settings::WARNING))
-                continue;
-            arithmeticError(tok,value);
+            const Token *operands[] = {tok->astOperand1(), tok->astOperand2()};
+            for(const Token **operand_it = begin(operands); operand_it != end(operands);operand_it++) {
```

Comment:
I am not sure about this code.

How about:

    for (int i = 1; i <= 2; i++) {
        const Token * const operand = (i==1) ? tok->astOperand1() : tok->astOperand2();

Or how about breaking out the for loop body in a function so no for loop is needed.

Reply:
Whats wrong with the regular for loop there?

Reply:
`std::begin` and `std::end` are not available in c++98. and I believe, unlike those functions you are not using an iterator.

Reply:
> std::begin and std::end are not available in c++98.

They are available as member functions in C++98. In C++11, they added free functions, so they can be used on arrays.

Also, I added `begin` and `end` function for arrays in utils.h, so they are available in C++98.

>  believe, unlike those functions you are not using an iterator.

A pointer meets all the requirements for a random access iterator. In fact, `std::iterator_traits` are specialized for pointers. This C++98 program prints out "Pointer is a random access iterator":

```cpp
#include <iterator>
#include <cstdio>

char is_random_access(std::random_access_iterator_tag);
long is_random_access(...);

int main() {
    if (sizeof(is_random_access(std::iterator_traits<int*>::iterator_category())) == 1)
        printf("Pointer is a random access iterator\n");
}

```

Reply:
sorry for late reply. I would like some kind of refactoring. I am skeptic about adding these `begin` and `end`.

I know that pointers can be used but I am not sure that it is specified they are used. For instance sometimes debug random access iterators implement extra runtime checks. access out of bounds etc. That can not be implemented in the library if it just uses a pointer typedef.

Reply:
How about skipping the for loop.

    const Token *pointerOperand;
    if (tok->astOperand1()->valueType() && tok->astOperand1()->valueType()->pointer>0)
        pointerOperand = tok->astOperand1();
    else if (tok->astOperand2()->valueType() && tok->astOperand2()->valueType()->pointer>0)
        pointerOperand = tok->astOperand2();
    else
        continue;

Reply:
> sorry for late reply. I would like some kind of refactoring. I am skeptic about adding these begin and end. I think that range for loops would be elegant, the day we start using c++11 I want to use range for loops here.

Well with PR #1138, I added a foreach emulation for older compilers so we can start using this now.

> I know that pointers can be used but I am not sure that it is specified they are used.

For `iterator_traits`, the standard library is required to provide the pointer [specialization](en.cppreference.com/w/cpp/iterator/iterator_traits#Specializations), and the `begin` and `end` array overloads are specified to return pointers as well(see [here](http://en.cppreference.com/w/cpp/iterator/begin)).

> For instance sometimes debug random access iterators implement extra runtime checks. access out of bounds etc. 

That is not a requirement for an iterator, and in fact, some older STL implementations used raw pointers directly. 

> That can not be implemented in the library if it just uses a pointer typedef.

Is it necessary for the iterators to `operands` have out of bounds check for debug? Accessing the array directly with `operands[i]` has no out of bounds check in debug, but I could add iterators with debug checking if you think its necessary.